### PR TITLE
Fix WebRequest failure to handle missing ContentType in response header

### PR DESCRIPTION
**Pull request by nmbradford**
If a response to invoke-webrequest does not set the content type in the
response header, an object not set exception is thrown in
WebResponseHelper.CoreClr.cs.  Updated to set contenttype = null.

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: Maybe `contentType = ContentHelper.GetContentType(response); `

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs]: Please remove unneeded empty line.

**Review by iSazonov — COMMENTED** [review]
@nmbradford  You would need to add tests.

**Comment by nmbradford** [comment]
I've added a single test for the exception that is thrown by iwr if Content-Type is missing from the response header and addressed the issues in review.  

Also - I have signed the CLA,. but the cla-required label is still attached to this issue?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Please remove unneeded empty lines L400 and L402.

**Comment by iSazonov** [comment]
@nmbradford Thanks! LGTM (Only correct the test). 
Wait for maintainers to review and remove cla.

**Comment by nmbradford** [comment]
done - thanks @iSazonov

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: Catch should be for a specific exception type

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs
Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs]: Same: catch should be for a specific exception type

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: Current comments are pretty vague and suggest that author of the code is not completely sure about what the code does. It's totally ok for a codereview discussion, maybe even test code, but it should be avoided in the product code.

Think about the comments from the next-person-who-is-going-to-edit-it perspective.
The purpose of the comments is to **clarify** some aspects of the code, not to admit that code may be written incorrectly.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: redundant, `contentType` is already assigned to null, on declaration

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs
Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs]: Same, redundent, because already assigned to null

**Review by vors — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (nmbradford) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: add explicit catch for NullReferenceException

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (nmbradford) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: Removed

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs
Comment (nmbradford) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs]: added explicit catch for NullReferenceException

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs
Comment (nmbradford) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs]: Removed as well

**Comment by nmbradford** [comment]
Thanks @vors - have made suggested changes...

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: NullReferenceException happens only when you are trying to de-reference null object. Exceptions are for exception situations and it's pretty easy to avoid NullReferenceException: we just need to check for `null` in the appropriate place. In this case, in `src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs` GetContentType.
There is a long chain of properties: `response.Content.Headers.ContentType.MediaType`. I don't know from the top of my head, which one is `null`, in this case, but I think you know.
Also that way we really fixing the problem in the implementation and not only on the caller side.

Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: That also means, that comment above should also be moved to the `ContentHelper.CoreClr.cs`

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs
Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs]: same

**Review by vors — CHANGES_REQUESTED** [review]

**Comment by nmbradford** [comment]
Thanks for the feedback @vors.  I've updated the PR to prevent the nullref instead of catching.  I also realised the same issue occurs in Invoke-RestMethod and the fix is almost identical, so I added a test for it and address it as well.  Let me know if this needs to go in a separate PR.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs
Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs]: C# 6 has a nice sugar for this, [Null-Conditional Operator](https://msdn.microsoft.com/en-us/magazine/dn802602.aspx)

```csharp
string charSet = response.Content.Headers.ContentType?.CharSet;
```

**Review by vors — CHANGES_REQUESTED** [review]
Thank you, it's great that you approach problem more generic and fixing `Invoke-RestMethod` at the same time!

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs
Comment (nmbradford) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs]: Ahh - that's nice and concise!  Thanks! Did you want me to update the PR with changes to that effect, or is just FYI (which is definitely appreciated since I'm a PowerShell guy who knows just enough csharp to hang himself ;) )

Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs]: Yes, can you, please, update the code?

Comment (nmbradford) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs]: done - back to a three character fix :)  Consider me schooled, thanks @vors

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs]: In the future it might not pass spelling verification. It would be nice to insert space after `//`
`// ContentType may not exist in response header.  Return null if not.`
And in below comments too.

Comment (nmbradford) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs]: Added.

**Comment by vors** [comment]
I updated the PR to remove extra blank lines

**Comment by vors** [comment]
CLA automation failed to remove `cla-required`, but CLA is signed.

**Comment by iSazonov** [comment]
@vors Maybe it is collision with @nmbradford mails?

**Comment by iSazonov** [comment]
@nmbradford Thanks for the successful completion ot the fix!

**Pull request merged**
